### ISSUE TITLE
Steam workshop support

### ISSUE THREAD

**Issue by phumberdroz** [Dahlgren/arma-server-web-admin#28]
As far as I can see there is currently no steam workshop support do you plan on adding it?

**Comment by Dahlgren** [comment]
Mission download is supported from the Missions page.

Addon and mod support is definitely something I'm planning to add but the Steam Web API does not seem to support it in the same way as for missions, no files are exposed using that.

I've been testing some alternative approaches to leverage this but I'm not yet 100% set on a good solution. Steamcmd seems to be the best way but then the mods end up in a steam directory with number ids as names instead of "normal" mod folder names so something needs to be done about that.

Management and overview about which mods are available, which are to be started with, download progress and updating needs to be clear and simple to use.

It's being looked into, it's something I'd really want since it would avoid being locked into withSIX and more users would probably be interested in using Steam Workshop anyway.

**Comment by flexgrip** [comment]
Dude. I just had a weird idea for this.

Was messing with steamcmd and their shitty directory structure after it downloads the workshop mods. Then I realized, every mod as a meta.cpp or a mod.cpp with contents that look like this:

`name = "Zade's BackpackonChest-Mod";
picture = "logo.paa";
actionName = "TF47";
action = "http://www.taskforce47.de";
description = "Zade's BackpackonChest-Mod - Version 0.21";
author = "DerZade";
tooltip = "Zade's BackpackonChest-Mod - Version 0.21";`

Maybe we can just parse these out and use the names to name the folders? The folder names really don't matter. It would be more about getting the UI to display their real names so users can perform tasks against them.

Anyway... cool project. Let me know if I can help.

**Comment by Dahlgren** [comment]
Yeah, the meta.cpp is definitely usable, good suggestion! Already have some parsers for Arma's class structures to use. The next hurdle is the lack of search functionality in the Steam Workshop APIs but that could be manual for now I guess, as with the missions download.

I'll see if I can integrate my steamcmd library with the web manager and hack together some basic mod management!

**Comment by Dahlgren** [comment]
Created a separate repo with all code related to Arma 3 Steam Workshop, https://github.com/Dahlgren/node-arma-steam-workshop

It can read which mods are available with their id and name from the meta.cpp (thanks @flexgrip!). Next step would be to implement steamcmd integration for Linux usage at least. Not sure if Windows could utilize Steam if installed or if it should always use steamcmd. Might be annoying if a mod is downloaded twice when Steam is installed and a mod is subscribed on the account.

**Comment by flexgrip** [comment]
What about getpublishedfiledetails() ?

I haven't used the web api. But is that the function you were saying doesn't seem to work for mods? The docs say that will give you the name, dates, meta data, etc.

**Comment by phumberdroz** [comment]
Will check out if I can help. 

There is also this: 

http://steamworkshopdownloader.com/

--
ha. Funny thing is the example mod is one of my Community.. :D

**Comment by Dahlgren** [comment]
`GetPublishedFileDetails` takes an array of workshop item ids and returns metadata about it such as title, last updated, description etc. Would be used to check if an item is updated. For missions published on Steam Workshop it also returns a download url but not for addons.

I've previously created a module for downloading metadata and mission files from Steam Workshop which is used by the web manager for mission downloads, https://github.com/Dahlgren/node-steam-workshop

**Comment by phumberdroz** [comment]
Apparently that is the way to download workshop stuff with steamcmd: 
/home/Steam/steamcmd.sh +login <user> <password> +workshop_download_item 107410 <ws_item_id>

**Comment by Dahlgren** [comment]
Yep, I have a WIP with workshop items management using the steamcmd. The main headache with steamcmd and workshop items is that it is not possible to force a directory as you can do with app installs. The mods will always end up in the steamcmd directory regardless of options, environment working directory etc.

**Comment by flexgrip** [comment]
You are right about the stupid steam directory thing. Even if you use force_directory /whatever it still puts its in `/path/to/steamcmd/steamapps/workshop/content/107410/`

However... what if we symlinked `ln -s /path/where/you/want/your/mods /path/to/steamcmd/steamapps/workshop/content/107410/`? Then it would download it to wherever you want, right?

**Comment by flexgrip** [comment]
Or... Even better, what if it kept the mods there, then after a mod is downloaded, we hit the web api to pull in the name. Then symlink `/path/to/steamcmd/steamapps/workshop/content/107410/812345 -> /mod/path/@mod_name`?

**Comment by Dahlgren** [comment]
The API does not return sane mod names, the latest X-Cam island is called `X-Cam-Taunus (Version 1.1)`. Of course that could be slugified etc but it would be weird to have a lot of folder just because they change the title. It's also good to avoid hitting the API too often.

Windows has directory junction which could work about the same as symlinks on linux but it's probably better not to mess around with the filesystem too much, keeping all mods in the steam folder also avoids bloating down the Arma directory.

All logic related to managing and querying the Steam Workshop downloads are already handled by the `arma-steam-workshop` library linked above. It returns all downloaded mods from steam workshop in a sane format with id, title and path abstracting away any complexity. Mods will appear in about the same was as they do today. Might want to split the local mods from the ones downloaded from Steam, might have to lists, not sure yet.

It will probably also be possible to migrate local mods that has a proper meta.cpp into the Steam Workshop mod format!

Integrating my steamcmd library with the new library remains as the next task :)

**Comment by flexgrip** [comment]
I think you're right. To be honest, the directory name doesn't even matter. The arma exe will launch a mod whether it's named @mod or 87655321. So you could just leave them in that folder and launch from wherever they are.

HOWEVER... This could be fixed, but we had another thread going on over at the Linux game servers Github page. Basically the problem is, on Linux, the box doesn't like mods unless they are in the same directory as the arma exe.

For example, arma3/mods/@mod_name worked. But mods outside of the arma3 root dir didn't work.

Maybe there's a workaround, I didn't put a lot of time into it.

All that to say, I hope you don't have to move the damned mods and manage all those paths.

**Comment by Dahlgren** [comment]
Thanks for the headsup, that needs to be looked into! Worst case I guess the steamcmd directory could be placed inside the Arma 3 folder

**Comment by flexgrip** [comment]
Yeah that could work. I gotta say... I was looking through the code. I'm surprised you could do so much with node. You are a braver man than I 👍 

I figured you'd at least mix some python or something in there :)

**Comment by Dahlgren** [comment]
Would it be okay to store steam login information in the config.js file?

Need some handling of steam guard too, but that needs to be done on demand. The current process handler does not trigger on steam guard requirement due to buffering of console output or something. I've been looking at some different terminal emulators to react to steamcmd output quicker.

Sadly it seems to only show progress for app updates and not workshop downloads. Might get Arma 3 updating through steamcmd for free at least when this is done!

**Comment by flexgrip** [comment]
Sure. I don't have any problems using my personal steam login on the server. Most of the time there's only one, maybe two admins. In my case, I'm the only one who even logs in to the server.

As far as the buffering, I know what you mean... I wrote some bash scripts to download workshop content recently. However, after you login the first time and enter the steam guard, it doesn't ask you for the code until you deauth the device. So maybe you can just set a timer or something... or maybe even pipe the output to some tmux session or something? I'm just shooting in the dark.

**Comment by flexgrip** [comment]
I just thought of a way you could get the steam guard code entered in an easy way. Ask the user to login with user/pass. Then tell them they need to check their email for the steamguard code and kill the steamcmd process. Then ask them for the steamguard code, and then log them in with +login user pass code. Could that work?

**Comment by Dahlgren** [comment]
I have not forgotten about this! I'm gonna focus on getting 1.68 support done first including x64 servers since it will probably ship to the stable within the next weeks, first RC was released yesterday.

**Comment by flexgrip** [comment]
Not trying to pressure you. Just wanted to see if there's anything we can do to help with this and if you have any sense on when you might be able to add this?

In really wanting to switch our unit out from swifty to the workshop. Which wouldn't be hard for the clients, but every time a mod gets updated I need to update the server manually :(

**Comment by Dahlgren** [comment]
I'm gonna finish the separation of Play withSIX integration from the core functions since Steam will work the same way. I think most of the Workshop handling works in the other module I created so it's probably more about making the steamcmd module behave better, especially regarding steam guard since currently it just hangs without any indication as to why.

**Comment by flexgrip** [comment]
What other module?

**Comment by flexgrip** [comment]
Oh you mean the module that grabs the workshop missions?

**Comment by Dahlgren** [comment]
https://github.com/Dahlgren/node-arma-steam-workshop and https://github.com/Dahlgren/node-steamcmd

**Comment by flexgrip** [comment]
How do I use that?

**Comment by flexgrip** [comment]
So, for the time being, I got workshop mods working. I subscribed to them from steam, symlinked them to the arma directory, and then added some options to the line `return file.charAt(0) == "@"` in mods.js so that it would read in directories that started with numbers, as workshop mods do.

This is still a fairly manual process, but it works reasonably well and I wanted to share some things I learned.

So now that I added the rudimentary ability for your app to see the workshop mods, I added some jquery on the mods page that looks at each workshop mod ID and uses the steam API to turn those ID's into a readable name. Like 12345678 = "cba_a3". So now the mod page shows the mod names instead of an unreadable number.

I'm using the API method `getpublishedfiledetails`. It spits back json with the title and the preview image and stuff.

The drawback about this hacky way of doing it is it doesn't track my subscribed workshop items. If I subscribe to a new mod I have to go manually symlink it to the arma path. So your implementation will be such a time saver in the future.

Also, about the steamcmd / steam guard issue. I tried doing steam family sharing to a second account, hoping that could be the server account and I could just share my arma to it and then use that to subscribe and get the mods through steam. However, I found out that arma is one of the few games you can't "family share" to another account.

However I did turn off steam guard, which seemed to make steamcmd work better. Before, I couldn't close steamcmd in a bash script. Now with steam guard off, it will exit normally.

Anyway, hope this helps. Love this app. Great work! This makes running our servers so much easier.

**Comment by Dahlgren** [comment]
The workshop library linked in the previous comment resolves mod title from the meta.cpp files so that no requests to Steam Workshop are necessary. It also finds all mods from the Steam Workshop directory so that mods don't have to be placed inside the actual Arma 3 folder.

Steam Guard Auth code can be passed with the option `+set_steam_guard_code <code>` to steamcmd

**Comment by flexgrip** [comment]
Hmm, so how would this all fit together? If that's what you're working on, cool. If not, I can try to get them all working together today.

**Comment by Dahlgren** [comment]
It will add additional mods to the list of mods or as a separate list above/below, I'm not sure what makes most sense :)

**Comment by flexgrip** [comment]
Well damn :) can I just npm install them or do I have to manually clone the source and put the files in arma-server-web-admin and require()?

**Comment by Dahlgren** [comment]
There's an early WIP which manages already installed Steam Workshop mods available in the branch `feature/steam-workshop-mods`. No installation or updating is available yet.

**Comment by ToasterBR** [comment]
@Dahlgren, just to let you know, our unit has been using the `feature/steam-workshop-mods` with success. 

There are some minor issues, although:
1- If the mod is set to private (or to friends), the `meta.cpp` sets the `publishedid = 0;`. 
This will break the fork, since it uses to identify the mods and list it in the web app.
Our workaround is to manually set the `publishedid` to the correct id of the mod.

2- In one case, we downloaded the new Lythium map by FFMOD team. It uses JBad as an requirement. Problem is that Jbad came without` meta.cpp`. Our work around was to create a new one and fill it with the correct data.
If you want to investigate this further, here are the link to JBad: http://steamcommunity.com/workshop/filedetails/?id=520618345

3- I believe your fix to correct the upload because of Express 4 was only done in the main branch. The error still exists in the  `feature/steam-workshop-mods` branch.

In the end, this was a HUGE time saver for us. We were using Swifty in the past, but to better play with other units, we changed to Steam Workshop. Without your app, I would give up. 

Thanks and send a donation URL.

Cheers, 
Toaster

**Comment by Dahlgren** [comment]
Oooh, cool, thanks for the feedback! :D

Really nice with the feedback on  `1`, I couldn't figure out why it was missing the id so I've added the mod folder name as fallback since that is/should be the mod id. Should probably publish that fix and update the web admin.

I've got reports on some other mods that is also missing the `meta.cpp` file. I'm thinking about adding the `mod.cpp` as fallback to fetch a nice title. I could also use the mod folder name id and lookup the metadata from the Steam Workshop API. Not sure what the best course of action is, maybe both since some mods might not have a `mod.cpp`.

I'll publish some updates and include the fix for `3` ASAP :)

We're discussing adding Steam Workshop as a source for our current Arma3Sync repository in addition to fetching mods from withSIX so hopefully the library will get more love soon.

**Comment by Dahlgren** [comment]
1 and 3 should be fixed on the updated branch now!

**Comment by Dahlgren** [comment]
Seems like steamcmd recently made some changes to how mods are handled regarding which folder they end up in when forcing a path

**Comment by Dahlgren** [comment]
Searching and Downloading should work somewhat now. Next in line is to keep track of current downloads to show in the UI (probably won't show progress) and proper cleanup on deletion. After that, checking for updates 🙃

**Comment by flexgrip** [comment]
Is that using steamcmd?

**Comment by Dahlgren** [comment]
Yep!

**Comment by flexgrip** [comment]
Just curious, since this is a node app, what method were you thinking of to get updating working through steamcmd? I'm assuming the node service can run some type of cron or queue?

**Comment by Dahlgren** [comment]
First step would be manual, showing the mod as outdated in the list. I could also add some kind of cronjob that can be enabled to autoupdate mods if desired.

**Comment by flexgrip** [comment]
Awesome. I will look through the current branch and see if I can help at all.

**Comment by flexgrip** [comment]
So updating a mod via steamcmd is the exact same steamcmd command as installing it, right?

**Comment by Dahlgren** [comment]
Yeah, updating the mod is implemented, checking for updates is not

**Comment by flexgrip** [comment]
Hey @Dahlgren - Any idea whats up with this on steam branch?

![](http://d.pr/i/KXhLLR+)

**Comment by Dahlgren** [comment]
Most often permission error, either for any of the steam data directories or workshop directory. Did you previously run anything as root/admin user?

Could also be out of disk space.

**Comment by flexgrip** [comment]
This is actually on my Windows server. I know a lot about permissions on *nix, but not on Windows. I've changed the windows service to run via my user name and the system user. Problem still persists. Maybe I'll see if I can just make everything globally rw in that steam folder. Where does steamcmd try to download the files to?

**Comment by Dahlgren** [comment]
The value specified in the config for steam -> path

**Comment by flexgrip** [comment]
I'll try moving that. When I first ran it, I saw it print something to stdout like, "Checking for steamcmd. ..... no steamcmd so downloading it." Paraphrasing.

**Comment by Dahlgren** [comment]
Steamcmd is saved to your users appdata (or similar)

**Comment by flexgrip** [comment]
Yeah I can't figure out why I'm getting that failed (I/O Operation Failed). I've looked at the permissions on this windows server for the steam directory, but I don't see what's up. Maybe the steamcmd that downloaded needs some special permissions?

**Comment by flexgrip** [comment]
Yeah trying my windows server again. If I use your node app to install mods I get I/O Operation Failed. If I go into that steamcmd directory under AppData and run steamcmd manually, it downloads files all day long and works great.

What gives? Is there any way to get additional logs out of it or something? I'm at a loss. I got the workshop branch to work well on linux. But Arma itself sucks on linux and gives people random crashes. So reverting back to windows is my only course of action.

**Comment by flexgrip** [comment]
Actually... it seems like it only happens when I force_install_dir to the steam directory. If I download it outside of their, it works fine. So something is stopping it.

I made sure steam wasn't running and nothing had a lock on the files. I just don't get it...

**Comment by Dahlgren** [comment]
Make sure your steam workshop path is correct and in a valid format for usage on Windows. Windows used backslash which needs to be escaped in the config file comparwd to Linux forwardslash. Since you mention `force_install_dir` it would seem the path is invalid.

**Comment by flexgrip** [comment]
You gotta be kidding me. That was it...

`path: 'C:\\Program Files (x86)\\Steam',`

Thanks bro. That should get be back up and running on windows.

**Comment by Dahlgren** [comment]
I think most of the Steam issues are resolved on the other WIP branch `feature/steam-workshop-mods-status`. It has checking for updates, updating, fixing corrupt/partial downloads and a different way for finding mods without relying on meta.cpp to exist.

Anyone tested it already? 🙂

**Comment by ToasterBR** [comment]
Gonna test ASAP! Thanks!
